### ISSUE TITLE
Find references: history needs polish

### ISSUE THREAD

**Issue by bpasero** [Microsoft/vscode-references-view#35]
At first when I saw this I thought the view was actually broken, maybe partially disposed and showing some raw HTML:

![image](https://user-images.githubusercontent.com/900690/49139105-3846bc00-f2f1-11e8-95b7-0d1ae3c9ddea.png)

I suggest to show a simple message when the view is empty, similar to other views, but not complex data. Maybe have a dropdown as action that shows recent queries.

//cc @joaomoreno @sandy081 (we talked about this last week already)

**Comment by jrieken** [comment]
@sandy081 I cannot control the colors from the extension...

**Comment by bpasero** [comment]
I think allowing for links etc. in messages asks for trouble, I would not allow markdown string in empty view messages at all.

**Comment by sandy081** [comment]
> I suggest to show a simple message when the view is empty, similar to other views, but not complex data. Maybe have a dropdown as action that shows recent queries.

Hence assigned to you.

@bpasero One reason to have markdown string is to have simple links to add entry (for eg: server/database) link in the message.

Created a separate issue for badly rendered links https://github.com/Microsoft/vscode/issues/63928

**Comment by jrieken** [comment]
> Hence assigned to you.

Missed that. Having the history (with links) in there was the whole point of the design and is the outcome of the UX call on this. I think it's works well, esp. given the lack of an input box or similar. I don't think that menu entries will be a viable alternative as these entries tend to be long'ish

**Comment by bpasero** [comment]
@jrieken @sandy081 what I don't like is that we use a tree/list for the results when finding references but when showing a list of history entries we use the most basic (and to be honest - ugly) HTML list rendering? What about users that want to navigate this list with the same keybindings as other widgets that show a list of elements (List,Tree)?

I actually question: a) the look and b) the usefulness of having such a history in the first place. Maybe I missed that UX call and if so, can we please put this on the UX agenda again?

**Comment by bpasero** [comment]
On top of that: I actually would not be happy to see other custom views show markdown strings as message with links. We are not doing this anywhere else. Empty views show a button to ask for an action (e.g. empty explorer) if any. But that is a discussion that can take place elsewhere.

**Comment by jrieken** [comment]
> We are not doing this anywhere else

We do exactly that in hovers - gitlens is the reference here and I think everyone likes that. I think there is value. You might argue the usefulness of a reference search history but I think many extensions benefit from that. The pattern is "explain how this view is populated" and it applies to all kinds of (azure, sql,...) resource trees that require some kind of login or config step. Today, extensions fake this by having a single item in the tree and I don't think that's great.

Styling is something that needs to be tweaked but supporting these kind of messages is good for better extensions. 

> the usefulness of having such a history in the first place

We have histories in many places (like search et al). The value it brings is that you can drill into references, go down the rabbit hole, and have a way to recover previous searches.

**Comment by jrieken** [comment]
Just to tame the excitement: there is no support for tables.

**Comment by jrieken** [comment]
> Just to tame the excitement: there is no support for tables.

Well, there actually might be ;-) Yeah, I agree that we should hand pick some MD features that we don't support. Not sure about being super minimal like notifications (for which I know people feel constraint) but I think people folks play nice and I take the various places that already allow MD strings as references. 

What I don't want to end up with is yet another type that says 'this is markdown but actually not'

**Comment by joaomoreno** [comment]
> Bottom focusable area is the empty tree. Will hide the tree when it is empty and there is a message (or may be always)

Make sure the message is focusable.

As for actions, I think reusing the menu concept would make more sense here. Inline actions as well as a proper dropdown menu would get the job done.

**Comment by jrieken** [comment]
> a proper dropdown menu would get the job done.

Can you sketch this up? How would a drop down menu for recebt reference searches look like?

**Comment by sandy081** [comment]
> but I think people folks play nice and I take the various places that already allow MD strings as references.

Well.. Happy if people are nice but cant depend on it. They just see that a feature is available and starts using it. This will abuse the Tree view by rendering completely a different view in tree view. I do not think Tree view and others (Hover) are similar in terms of using MD string because the latter's main data source is MD and not for Treeview.

Regarding history do we do the same, if want to support it in Search?

**Comment by bpasero** [comment]
I think history in search works nicely because of the input field, if something like that would exist for references, it could work the same way.

**Comment by jrieken** [comment]
> if something like that would exist for references, it could work the same way.

Like how? Do you wanna type reference command? It's is really different from search. Also, the dropdown menu options (for which we have no technical support btw) don't fly. An alternative, albeit much more complicated and less user friendly, is a command that open quick pick with previous searches...

**Comment by bpasero** [comment]
@jrieken it does not have to be an input field that is editable, it could be some kind of widget that shows the "thing" you are finding references for and then have a little triangle dropdown to pick previous searches. In a way it would probably behave like a dropdown similar to the debug one:

![image](https://user-images.githubusercontent.com/900690/49220249-05c2bf00-f3d6-11e8-97d7-568eda54fdea.png)

**Comment by jrieken** [comment]
> little triangle dropdown to pick previous searches. In a way it would probably behave like a dropdown similar to the debug one:

The different is that debug targets have names and that reference search is invoked out of some source code and I think showing that in these kind of UIs is problematic

**Comment by bpasero** [comment]
Not sure, maybe a label can be found for a reference based on its name and container. As comparison, Eclipse has a label to the top and some more details, like the number of references found which could also be useful to know:

![image](https://user-images.githubusercontent.com/900690/49237943-ae394900-f3ff-11e8-890b-1d56b6971d88.png)

**Comment by jrieken** [comment]
> like the number of references found which could also be useful to know:

Dude, open your eyes. 

> As comparison, Eclipse has a label to the top and some more details,

That's something we don't have because you talk locations and not symbols.

**Comment by bpasero** [comment]
> Dude, open your eyes.
?!, just copy what we do in find references when embedded, I see no reason why I would not want to see the same info there:

![image](https://user-images.githubusercontent.com/900690/49288025-d6788480-f49f-11e8-9811-069034c4439f.png)

**Comment by bpasero** [comment]
@jrieken I still do not understand why you cannot use:
* the message about the view being empty and that you can click on an item below to populate the view from a previous search
* a tree with previous searches

Similar to what Gitlens is doing so far to help users to populate a view. Given that we would only need the message to be a normal string and everything else is already there.

**Comment by jrieken** [comment]
The question is if that's better? The single, big button isn't something I would associate with a history. Then my very first implementation was actually having a tree node per search, with the actual results has children. It introduces lots of UI clutter because you don't have loads of history entries

**Comment by bpasero** [comment]
@jrieken yes, a single button might not be easy to associate with a history. Having a tree with previous searches that includes the elements might be too much, I was more referring to something like https://github.com/Microsoft/vscode/issues/61312#issuecomment-439105876 where you have a message "Previous find references searches:" and then below a flat list of searches (like you do today with the HTML message). Clicking on any entry will execute the search.
